Fix busy_sleep_micros and ibex_busy_loop

alphan opened the pull request:
While I was rebasing my changes for the GPIO DIF, I noticed that
bouncing LEDs part of `hello_world.c` did not work correctly. This
change fixes the following:

- `busy_sleep_micros(size_t)`:
  - Modified to use a `uint64_t` to store the number of cycles instead
of `uint32_t`.
  - Renamed to `usleep(uint32_t)`.
- `ibex_busy_loop(size_t)`
  - Added code to make sure that `cycles` is a multiple of 8.
  - Modified to expect a `uint64_t` instead of a `uint32_t`.
  - Modified inline assembly to work with a `uint64_t` (split into two
`uint32_t`s). Since this is my first time writing inline asm in this
project, I would appreciate if you could review this part more closely.

See also: #1227, #1207.

Note: I reverted my earlier change that moves division before
multiplication in `udelay(uint32_t)` because `kIbexClockFreqHz` is
defined as 50 kHZ for simulations in `ibex.c`.

Signed-off-by: Alphan Ulusoy <alphan@google.com>

lenary commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -33,7 +33,7 @@ inline void ibex_busy_loop(size_t cycles) {
       "  nop; nop; nop; nop;"    // 4 cycles.
       "  addi %1, %1, -8;"       // 1 cycle.
-      "  blez %1, busy_loop%=;"  // 3 cycles.
+      "  bgez %1, busy_loop%=;"  // 3 cycles.
```
Oops, I think I was the person that said it should be `blez`. Sorry for the confusion.

alphan replied:
No worries and thanks for the clarification about the data types on rv32!

gkelly commented:
I'm confused about the Azure failure.  It's trying to check out a commit that's not this commit, after successfully fetching the tree.  The commit it's _trying_ to check out doesn't appear to be real, it's neither of the parents for the actual merge head...

lenary commented:
I think it is trying to checkout the commit that represents if this PR was merged into master, right now. I have just asked it to retry the jobs, which should work.

imphil commented on sw/device/lib/runtime/hart.c:
```diff
@@ -11,7 +11,7 @@
 extern void wait_for_interrupt(void);
 
 void busy_sleep_micros(size_t microseconds) {
```
This is a strange name and data type for a function commonly called `usleep(unsigned int usec)`, so if you want to change the data type that's fine with me.

alphan replied:
I agree. I will update this PR.

mcy replied:
I believe I renamed it from `usleep` because that function tends to have connotations of yielding to a scheduler. I think a spin-sleep should be marked as such.

Also, as Sam said, you want a larger data type, though you should prefer to spell it with a stdint.h name, like `uint64_t`.

lenary commented:
For the record, `unsigned long` is 32 bits (the same size as `unsigned int`) on rv32. You will want `unsigned long long` if you want to extend the data type.

mcy commented on sw/device/lib/runtime/hart.c:
```diff
@@ -11,7 +11,7 @@
 void busy_sleep_micros(size_t microseconds) {
-  size_t cycles = kIbexClockFreqHz * microseconds / 10000000;
+  size_t cycles = kIbexClockFreqHz / 1000000 * microseconds;
   ibex_busy_loop(cycles);
```
I think I'd make this int64_t arithmetic (and mark the literal `LL`) for overflow/underflow protection.

alphan commented:
(@imphil, @lenary)
- `busy_sleep_micros(size_t)`
  - Rename to `udelay(uint32_t)`. 
  - Use a `uint64_t` to store the number of cycles instead of `uint32_t`.
- `ibex_busy_loop(size_t)`:
  - Added code to make sure that `cycles` is a multiple of 8.
  - Modified to expect a `uint64_t` instead of a `uint32_t`.

imphil commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -27,15 +27,28 @@ extern const size_t kIbexClockFreqHz;
+  uint32_t cycles_upper = cycles >> 32;
+  // To tell the compiler that |cycles_lower| and |cycles_upper| are modified.
+  uint32_t out_lower, out_upper;
+  // Takes approximately 8 cycles per interation
```
Shouldn't it be exactly 8 cycles?

alphan replied:
Ideally, yes. I wrote _approximately_ because of the outer loop (labelled `upper`), which runs once per `UINT32_MAX` cycles. I can add more details or remove the comment if it is confusing/misleading.

alphan replied:
Using the cycle counter resolves this issue.

imphil commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -27,15 +27,28 @@ extern const size_t kIbexClockFreqHz;
+      "  beqz %3, done%=;"   //   Branch to |done| if |cycles_upper| is zero.
+      "  addi %2, %2, -8;"   //   Decrement |cycles_lower| by eight.
+      "  addi %3, %3, -1;"   //   Decrement |cycles_upper| by one, i.e. borrow.
+      "  j lower%=;"         //   Jump to |lower|.
```
Does this really add up to 8 cycles? Please have a look at the waveform to confirm.

GregAC replied:
I would note the stall cycles on the branch and jump does depend upon the interconnect. If I remember correctly this is actually different between the FPGA and Verilator versions of earlgrey.

It may be better to simply spin reading mcycle until it hits the appropriate value?

mcy replied:
I would be in favor of switching to mcycle. It would make this code not be Ibex-specific (and we can, as such, just delete `ibex.h`, once my link-time PR goes through).

alphan replied:
Thank you @imphil for looking at the asm closely and @GregAC for pointing this out! I actually verified the behavior using a stopwatch before sending the PR (yes, I wasn't aware that verilator could produce a core trace, still ramping up) and it looked correct. But, I see that the core trace from verilator shows that this snippet takes 4 cycles. So, I added 4 `nop`s (second column is cycle):
```
         177946	     88968	200001a6	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177948	     88969	200001a8	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177950	     88970	200001aa	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177952	     88971	200001ac	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177954	     88972	200001ae	c119	c.beqz	x10,200001b4	 x10:0x0007a120
         177956	     88973	200001b0	1561	c.addi	x10,-8	 x10:0x0007a120 x10=0x0007a118
         177960	     88975	200001b2	bfd5	c.j	200001a6	
         177962	     88976	200001a6	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177964	     88977	200001a8	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177966	     88978	200001aa	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177968	     88979	200001ac	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177970	     88980	200001ae	c119	c.beqz	x10,200001b4	 x10:0x0007a118
         177972	     88981	200001b0	1561	c.addi	x10,-8	 x10:0x0007a118 x10=0x0007a110
         177976	     88983	200001b2	bfd5	c.j	200001a6	
```
But, when I run it on the FPGA, it is slower than it should be (~1/3). As far as I can see, RTL and C implementation are in agreement about the clock frequency (50 MHz) so this is probably due to what @GregAC pointed out.

I like the idea of using `mcycle` since it eliminates the need for hand-tuned instructions. I need to look it up.

alphan replied:
Using the cycle counter resolves this issue.

imphil commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -27,15 +27,28 @@ extern const size_t kIbexClockFreqHz;
  */
-inline void ibex_busy_loop(size_t cycles) {
-  size_t out;  // Used to create an inout parameter below.
+inline void ibex_busy_loop(uint64_t cycles) {
```
I can't comment further up in the docstring, but typical behavior in `usleep()` and friends is to spin for *at least* the amount of cycles.

alphan replied:
Using the cycle counter resolves this issue.

imphil commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -27,15 +27,28 @@ extern const size_t kIbexClockFreqHz;
-inline void ibex_busy_loop(size_t cycles) {
-  size_t out;  // Used to create an inout parameter below.
+inline void ibex_busy_loop(uint64_t cycles) {
+  // Round down |cycles| to closest multiple of 8
```
If we want the behavior of spinning at least for `cycles` cycles, then this needs to round up.

alphan replied:
Sure, I rounded it down to keep things simple (since rounding up would require range checking to prevent overflow) but I will change this part as you suggested.

alphan replied:
Using the cycle counter also resolved this issue.

alphan commented:
- Using `usleep()` instead of `udelay()`.
- Rounding up instead of down.
- Updated asm.

Next: `mcycle`

mcy commented:
I'm a bit comme-si comme-ça about `usleep()`, mostly because it implies descheduling or low power mode (and I'm generally not in favor of re-using Unix names for stuff because of highly overloaded meaning). I'm not going to block this change on that, though.

alphan commented:
- Use cycle counter instead of counting down in `asm`.

imphil commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -20,22 +20,29 @@
  */
-inline void ibex_busy_loop(size_t cycles) {
-  size_t out;  // Used to create an inout parameter below.
+inline uint64_t read_cycle_counter() {
```
- This isn't really Ibex-specific, so it probably can go into another file if you'd like. But fine with me either way.
- I'd also think about a name which sounds a bit more like a getter, like `get_mcycle()`. But again, up to you.

alphan replied:
Renamed to `ibex_mcycle_read()`.

imphil commented on sw/device/lib/runtime/hart.c:
```diff
@@ -10,9 +10,11 @@
+void usleep(uint32_t usec) {
+  uint64_t cycles = (uint64_t)kIbexClockFreqHz * usec / 1000000;
+  uint64_t start = read_cycle_counter();
+  while ((uint64_t)(read_cycle_counter() - start) < cycles) {
```
Isn't that cast unnecessary?

alphan replied:
Yes, (in this case, I think, not if we used something smaller like `uint16_t`) but I can't always remember all the integer promotion rules so I use the cast to be safe. I can definitely remove it if you want.

alphan replied:
Removed.

imphil commented on sw/device/lib/runtime/hart.h:
```diff
@@ -26,9 +26,9 @@ inline void wait_for_interrupt(void) { asm volatile("wfi"); }
 /**
  * Spin for roughly the given number of microseconds.
```
roughly -> at least

alphan replied:
Done, thanks!

gkelly commented on sw/device/lib/runtime/ibex.h:
```diff
@@ -20,22 +20,29 @@
  */
-inline void ibex_busy_loop(size_t cycles) {
-  size_t out;  // Used to create an inout parameter below.
+inline uint64_t read_cycle_counter() {
```
I think this should be `ibex_cycle_counter_read`. I know it's not ibex-specific, but `ibex.h` is where this kind of stuff is going for now.

gkelly replied:
`ibex_mcycle_read`, @imphil thoughts?

imphil replied:
Sure, sounds good.

alphan replied:
Done.

alphan commented:
- Renamed `read_cycle_counter()` to `ibex_mcycle_read()`.

imphil commented:
@alphan ready to merge?

alphan commented:
> @alphan ready to merge?

Yes, thanks!
